### ISSUE TITLE
GTK4

### ISSUE THREAD

**Issue by LukashonakV** [Alexays/Waybar#2815]
Hi @Alexays , @alebastr  I've done some digging into GTK4 and opportunity to migrate Waybar to it. 
So. I have the first base example of the Waybar which can be build and run with gtkmm4 . 
There only base engine... without any modules and event handlers. But I think it's a good start to migrate Waybar. 
@Alexays  is it possible for you to create for example separate branch and mark it with labels: unstable, gtk4 and to do cloning from my branch : https://github.com/LukashonakV/Waybar/tree/gtk4
I think it will provide an opportunity to contributors to make contributions, fix errors, find more effective approaches, bring new ideas, speedup the process

About versioning: 4.1.0  - the first digit means (stable release version. Currently it points to GTK4 chain), the second digit experimental version. The third digit update of the stable release. So for example the first stable release of the Waybar GTK4 should be looked like: 4.0.1. - The same versioning system now for example is used by Wine project.
So current 4.1.0 - meanse - Waybar 4 experimental 1 release without updates of the stable release

Once this approach is approved and separate branch is created it's possible to keep TODO list in this issue.

Thank you

----
Migration tasks
- [x] [Backlight](https://github.com/Alexays/Waybar/wiki/Module:-Backlight)
- [x] [Backlight/Slider](https://github.com/Alexays/Waybar/wiki/Module:-Backlight-Slider)
- [x] [Battery](https://github.com/Alexays/Waybar/wiki/Module:-Battery)
- [x] [Bluetooth](https://github.com/Alexays/Waybar/wiki/Module:-Bluetooth)
- [x] [Cava](https://github.com/Alexays/Waybar/wiki/Module:-Cava)
- [x] [CFFI]()
- [x] [Clock](https://github.com/Alexays/Waybar/wiki/Module:-Clock)
- [x] [CPU](https://github.com/Alexays/Waybar/wiki/Module:-CPU)
- [x] [Custom](https://github.com/Alexays/Waybar/wiki/Module:-Custom)
- [x] [Disk](https://github.com/Alexays/Waybar/wiki/Module:-Disk)
- [x] [DWL](https://github.com/Alexays/Waybar/wiki/Module:-Dwl)
- [x] [Gamemode](https://github.com/Alexays/Waybar/wiki/Module:-Gamemode)
- [x] [Group](https://github.com/Alexays/Waybar/wiki/Module:-Group)
- [x] [Hyprland](https://github.com/Alexays/Waybar/wiki/Module:-Hyprland)
- [x] [Idle Inhibitor](https://github.com/Alexays/Waybar/wiki/Module:-Idle-Inhibitor)
- [x] [Image](https://github.com/Alexays/Waybar/wiki/Module:-Image)
- [x] [JACK](https://github.com/Alexays/Waybar/wiki/Module:-JACK)
- [x] [Keyboard State](https://github.com/Alexays/Waybar/wiki/Module:-Keyboard-State)
- [x] [Language](https://github.com/Alexays/Waybar/wiki/Module:-Language)
- [x] [Memory](https://github.com/Alexays/Waybar/wiki/Module:-Memory)
- [x] [MPD](https://github.com/Alexays/Waybar/wiki/Module:-MPD) No guarantee
- [x] [MPRIS](https://github.com/Alexays/Waybar/wiki/Module:-MPRIS) No guarantee
- [x] [Network](https://github.com/Alexays/Waybar/wiki/Module:-Network)
- [x] [Niri](https://github.com/Alexays/Waybar/wiki/Module:-Niri)
- [x] [PulseAudio](https://github.com/Alexays/Waybar/wiki/Module:-PulseAudio)
- [x] [PulseAudio/Slider](https://github.com/Alexays/Waybar/wiki/Module:-PulseAudio-Slider)
- [x] [Privacy](https://github.com/Alexays/Waybar/wiki/Module:-Privacy)
- [x] [Sndio](https://github.com/Alexays/Waybar/wiki/Module:-Sndio) - need to check
- [x] [Systemd failed units](https://github.com/Alexays/Waybar/wiki/Module:-Systemd-failed-units)
- [x] [River](https://github.com/Alexays/Waybar/wiki/Module:-River)
- [x] [Sway](https://github.com/Alexays/Waybar/wiki/Module:-Sway)
- [x] WLR
- [x] [Temperature](https://github.com/Alexays/Waybar/wiki/Module:-Temperature)
- [ ] [Tray](https://github.com/Alexays/Waybar/wiki/Module:-Tray) - libdbusmenu is still in gtk3. Need to find another solution
- [x] [Taskbar](https://github.com/Alexays/Waybar/wiki/Module:-Taskbar)
- [x] [UPower](https://github.com/Alexays/Waybar/wiki/Module:-UPower) - fully rewritten
- [x] [WirePlumber](https://github.com/Alexays/Waybar/wiki/Module:-WirePlumber) No guarantee
- [x] [Workspaces](https://github.com/Alexays/Waybar/wiki/Module:-Workspaces)
- [x] [User](https://github.com/Alexays/Waybar/wiki/Module:-User)

----
Once Waybar is migrated need to do additional sync with the master in order to pick up all PR are done since migration process was going on

----
Dependencies:
- [x] [gtk4: missbehavior of the Gtk::Label tooltip](https://gitlab.gnome.org/GNOME/gtk/-/issues/6674)
- [x] [gtk3 migration gtk4 Label](https://discourse.gnome.org/t/gtk4-migration-label-and-tooltip-from-gtk3/20710)
- [x] [gtk4 Label](https://discourse.gnome.org/t/gtk4-migration-label-and-tooltip-from-gtk3-again/23999/2)
**Confirmed gtk4 - 4.17.0**

**Comment by alebastr** [comment]
First of all, I'm not really sure that's a good idea right now. The
approach taken in gtk4-layer-shell, MITM shim library that patches Wayland
protocol messages to replace xdg-shell with layer-shell, seems a bit
fragile to me. Any request or extension protocol that references
non-existing xdg-shell objects may cause the compositor to drop the client.
I'd personally stay away from GTK4 until
https://gitlab.gnome.org/GNOME/gtk/-/issues/2132 is resolved.
Although gtk4-layer-shell works with the current release of gtk4, so maybe
I'm too skeptical. At least it's worth experimenting with.

GTK4 migration will make gtk4-layer-shell mandatory. You'll need to drop
RawSurfaceImpl (it's not possible to make it work), and remove the meson
feature, the configuration flag and all the ifdefs.
This could be started in the current GTK3 version, as I don't see much
sense in keeping an inferior RawSurfaceImpl anymore. gtk-layer-shell is
widespread and I don't believe anyone's still building Waybar without
it. @Alexays,
WDYT?

Please, also check the attempt to replace GtkEventBox with GtkButton and
the resulting fallout. That would be another prerequisite for migration, as
GTK4 removed EventBox
<https://docs.gtk.org/gtk4/migrating-3to4.html#stop-using-gtkeventbox>, so
it might be worth recovering the code and addressing the remaining blocking
issues.

As a side-note, what are the benefits of GTK4 port? GTK3 will be maintained
for a while, since even GNOME haven't finished migrating all the
components. And by the time GTK3 is EOL, we'll have to start considering
freshly released GTK5 (which is not even being developed right now) :)

**Comment by LukashonakV** [comment]
Hi @alebastr , thank you for your response. This is definitely what I was needed , 
So. 
Regarding benefits... actually there are no any strong advantages waybar can get from the migration process... but according migration guide gtk team recommends to do migration from one major version to another... without any gaps. Of course it's not strictly.. but let's imagine tomorrow we will have GTK5 :))) In such case it would be more difficult to migrate Waybar from GTK3 to GTK5 ... + As I see GTK4 brought much more changes than for example GTK2->GTK3. ... So if it's possible such point can be counted as a one major benefit..
Second regarding eventbox . Actually as I saw according migration guide eventbox was dropped due to all items are are derived from the widget class . And Widget can handle any signals we need. But I still haven't had a chance to check it. If it's true I see we should revise a little AModule and ALabel architecture to satisfy new needs
Third... regarding this https://gitlab.gnome.org/GNOME/gtk/-/issues/2132 - as I see this it the long long long story :))) I think in cas if we wait this ticket is solved.... we will get GTK5 :)))
Actually as I see GTK team strategy is : If possible to avoid responsibility for interaction with Wayland. And seems like that's why the cutoff some APIS. Please see comment -> https://gitlab.gnome.org/GNOME/gtk/-/issues/2132#note_760523 . So I don't see situation will change in a good way :)

Please make sure I don't insist on migration  :) Just to try to figure out whether this good thing to start migration GTK3-GTK4... 
And please proceed with the sound criticism
Thanks

**Comment by Alexays** [comment]
Hi @alebastr and @LukashonakV.
I agree with Aleksei, I don't think anyone is still building Waybar without gtk-layer-shell, so I'm pretty ok with dropping RawSurfaceImpl to have only gtk-layer-shell.
But in view of a possible migration to GTK4, it would be a good idea to first drop RawSurfaceImpl and then check that on GTK4 the basic Widget class handles all the events, or whether we should try to migrate to GtkButton again.
Btw, I'll create a branch with your work to let anyone start working on it, with no guarantee that it will be merged into master in the near future.
To have a real advantage with this migration, it would have to bring new features that would be significant for users, why not listing these possibilities in this issue?

**Comment by smsteward** [comment]
@LukashonakV I had started doing this same thing a few weeks ago as a side project, albeit in Python (I don't know any C++ unfortunately due to some bad experiences with it in undergrad). I have some basic functions written and some template classes for the bar itself, but I'm having a hard time getting the widget templates off the ground (I'm a full time SQL dev, so my free time is quite limited). If you know any Python and would like a collaborator, I'd be down to help realize this.

**Comment by 5p4r74cu5** [comment]
> To have a real advantage with this migration, it would have to bring new features that would be significant for users, why not listing these possibilities in this issue?

@Alexays 

Hey :-) I apologise in advance if I am misunderstanding, as my technical knowledge is limited to bash scripting, but I've read in [other issues](https://github.com/Alexays/Waybar/issues/1838) that GTK4 may be relevant to implementing some aesthetically oriented features, like hover states.

I know that at the end of the day aesthetics will always be trumped by other more practical concerns, however basic features like hover states probably have a significant impact on peoples interest in using waybar. I know this is subjective, but it certainly gives the bar a somewhat dated feel.

**Comment by lucyColeclough** [comment]
@5p4r74cu5 
The main changes in gtk4 were to the development experience imo, with event controllers, not having to show every widget, snapshot structure, nuking cell renderers. Things like this.
The aesthetic differences were ported to gtk3 and waybar uses heavy css.
The main reason someone might want to keep up is the same with most projects in that the current version is the one that people are putting effort in to keep it compliant with all surrounding software. That said, people seem good at making old software still function, such as gimp which is still on gtk2.

**Comment by 5p4r74cu5** [comment]
@lucyColeclough 

Interesting, guess that explains why I gimp is so damn ugly.

Perhaps GTK4 has more of an indirect role in effecting the aesthetics, like regarding hover states, its mentioned as one of the things holding up #1120. I know its just one feature, but I think the hover effect has a disproportional impact of the user experience using waybar. States changing visually on hover is so typical of modern GUIs that it essentially feels like Waybar is broken, on an intuitive level.

That said, I hear you in that from a developer perspective the changes in the development experience will overshadow the effects for the average user, like me :-)

**Comment by lucyColeclough** [comment]
@5p4r74cu5 
Ah right forgot, yh in gtk3 you have to wrap stuff in an event box

**Comment by 4194304** [comment]
I fully support this. Being able to not use things like blur effects for me isn't acceptable. For now I'll be using workarounds to that specific problem

**Comment by LukashonakV** [comment]
Hi @Alexays , finally .... GTK4 branch more or less working well on my machine, 
Can you please force push gtk4 branch with the https://github.com/LukashonakV/Waybar/tree/gtk4 . 
What is done:
- [x] code base is using gtk4 framework
- [x] all commits from the master branch till dacecb9 are merged into gtk4 . So for now gtk4 branch is synced with the master
- [x] Linter is satisfied

What next:
1. Need to migrate docker files and update necessary dependencies to satisfy gtk4 branch
2. Need to migrated modules are left:
   - [ ] [Hyprland](https://github.com/Alexays/Waybar/wiki/Module:-Hyprland)
   - [ ] [River](https://github.com/Alexays/Waybar/wiki/Module:-River)
   - [ ] [Tray](https://github.com/Alexays/Waybar/wiki/Module:-Tray) - libdbusmenu is still in gtk3. Need to find another solution
   - [ ] [Taskbar](https://github.com/Alexays/Waybar/wiki/Module:-Taskbar)
   - [ ] [Niri](https://github.com/Alexays/Waybar/wiki/Module:-Niri)
 3. Rotation of the Label is deprecated in GTK4. So for now this option is disabled. Possible solution is using Cairo for that purpose. It needs to be investigated

----
@Alexays  can you publish this build as beta gtk4. ? Actually it's not an easy to merge such huge list of commits which where done since I started migration process. 

Thank you

**Comment by LukashonakV** [comment]
Hi @Alexays , GTK4 base migration routine is done. All modules were migrated. All commits from the GTK3 master were picked up. Please give some feedback )

**Comment by System64fumo** [comment]
@LukashonakV Just happened to stumble here and noticed you guys were looking for a tray alternative?
I've written a "somewhat functional" tray for my gtk4 bar [here](https://github.com/System64fumo/sysbar/blob/main/src/modules/tray.cpp)
Feel free to use that as a reference, Hope that helps.

**Comment by LukashonakV** [comment]
> Hi @Alexays , finally .... GTK4 branch more or less working well on my machine, Can you please force push gtk4 branch with the https://github.com/LukashonakV/Waybar/tree/gtk4 . What is done:
> 
>     * [x]  code base is using gtk4 framework
> 
>     * [x]  all commits from the master branch till [5f26051](https://github.com/Alexays/Waybar/commit/5f260516af9a3e5791c26c7fc18a85a47da5b23f) are merged into gtk4 . So for now gtk4 branch is synced with the master
> 
>     * [x]  Linter is satisfied
> 
> 
> What next:
> 
>     1. Need to migrate docker files and update necessary dependencies to satisfy gtk4 branch
> 
>     2. Need to migrate modules are left:
>        
>        * [x]  [Hyprland](https://github.com/Alexays/Waybar/wiki/Module:-Hyprland)
>        * [x]  WLR
>        * [x]  [River](https://github.com/Alexays/Waybar/wiki/Module:-River)
>        * [ ]  [Tray](https://github.com/Alexays/Waybar/wiki/Module:-Tray) - libdbusmenu is still in gtk3. Need to find another solution
>        * [x]  [Taskbar](https://github.com/Alexays/Waybar/wiki/Module:-Taskbar)
>        * [x]  [Niri](https://github.com/Alexays/Waybar/wiki/Module:-Niri)
> 
>     3. Rotation of the Label is deprecated in GTK4. So for now this option is disabled. Possible solution is using Cairo for that purpose. It needs to be investigated
> 
> 
> @Alexays can you publish this build as beta gtk4. ? Actually it's not an easy to merge such huge list of commits which where done since I started migration process.
> 
> Thank you

Hi @Alexays , any feedback ?

**Comment by Beryesa** [comment]
Could it be merged with/skipping/except the libdbusmenu as experimental/beta until that part is figured out instead of holding it here?

**Comment by apiraino** [comment]
> Could it be merged with/skipping/except the libdbusmenu as experimental/beta until that part is figured out instead of holding it here?

I - for one - find the Waybar tray a useful feature that I use, so I'd prefer not.

**Comment by apiraino** [comment]
@LukashonakV first of all, thanks for working on this. Some initial feedback on the GTK4 patch.

I've built Waybar (https://github.com/Alexays/Waybar/commit/86451156190e69afafb2162fb5891c0fc68c26e7) applying your GTK4 patch, I took note that the updated dependencies are `libgtkmm-4.0` and `libgtk4-layer-shell` (didn't identify more). When compiling the GTK4 branch I see some compile warnings, can't say if they were present in the GTK3 branch.

When running Waybar/GTK4 I've found a few issues. Where can I report them, where do you want to track them?
- "bar configured" info message repeats continuously
```
[2024-11-16 12:16:17.449] [info] Bar configured (width: 1920, height: 33) for output: eDP-1
[2024-11-16 12:16:17.481] [info] Bar configured (width: 1920, height: 33) for output: HDMI-A-2
...
```
- bluetooth module does not work (unsure why, nothing is logged on stderr/stdout)
- Tray icons do not appear (known issue)


Thoughts:
- Would it make sense opening a pull request on this repository to be sure your branch stays in sync with upstream and compiles cleanly?
- I wonder if the situation regarding gtk layer shell changed since this [comment](https://github.com/Alexays/Waybar/issues/2815#issuecomment-1888695407). Unsure about the context and implcitations here, but I could compile the Waybar+GTK4 patch by updating the `libgtk4-layer-shell` dependency.

**Comment by LukashonakV** [comment]
Hi @apiraino , thank you for your feedback.

So regarding some thoughts here

1. gtk4 should be at least 4.16 version. [gtk4 Label](https://discourse.gnome.org/t/gtk4-migration-label-and-tooltip-from-gtk3-again/23999/2)
2. Actually Alex already created gtk4 branch 10+- months ago when I have the first attempt to build waybar with gtk4. But honestly saying it was very raw variant. A lot of changes are made since that time. + Actually I've recreated base gtk4 branch in order to have tree of commits synced with the master branch. That's the only reason I cannot create pull request for that. If Alex just override gtk4 branch forced by mine it's just align not only code base but tree of commits to. Then it'll be much easy to backport incoming commits from the master branch to the gtk4 and vice versa.
3. Regarding bluetooth module. Let me check once I have time
4. Regarding workspaces... which module do you use hyprland/sway ? There is difference between their implementation
5. gtk branch has got gtk-layer-shell dependency version 4. https://github.com/LukashonakV/Waybar/blob/654a3268983c05740d6f5dbde1f117a1b731ccc8/meson.build#L108 . If the question about dependency naming... it can be solved easily ... if it's necessary :)

**Comment by apiraino** [comment]
I've investigated a little bit the issues I mentioned before, here are some minimal configs to reproduce:

* "bar configured" info message repeats continuously:
```
{
    "modules-left": ["network"],
    "network": {
        "interval": 1
    }
}
```

* workspaces and bluetooth modules not showing up. you can probably trigger this If you uncomment `fixed-center`
```
{
    "height": 33,
    //"fixed-center": false,
    "modules-left": ["sway/workspaces", "sway/mode", "bluetooth"]
}
```

Hope it helps, thanks!

**Comment by LukashonakV** [comment]
Hi @apiraino 
Actually the only thing I can reproduce is "bar configured" info message repeats continuously
All others stuff works for me well.
In the screenshot Top bar is running in GTK4. As you can see sway/workspaces in the left and bluetooth in the center works well for me. It might  be the reason that you build waybar using patch from the mine branch and some things are not took into the account.
![ps_2024-11-16-17_27_27](https://github.com/user-attachments/assets/21ba337f-9af8-45c7-b96d-556ef5ddeecb)

**Comment by apiraino** [comment]
@LukashonakV w.r.t. "workspaces and bluetooth modules not showing up". I can reproduce also with your branch. 

Just to be sure I made it clear how to reproduce, this config **will reproduce** the issue:
```
{
    "height": 33,
    "fixed-center": false,
    "modules-left": ["sway/workspaces", "sway/mode", "bluetooth"]
}
```
I also see a warning when starting Waybar:
```
(waybar:1542750): Gdk-WARNING **: 20:41:18.987: (../../../gdk/wayland/gdktoplevel-wayland.c:399):gdk_wayland_toplevel_compute_size: runtime check failed: (size.width > 0)
```

This other config **does not reproduce** the issue (and I don't see the warning either):
```
{
    "height": 33,
    // "fixed-center": false,
    "modules-left": ["sway/workspaces", "sway/mode", "bluetooth"]
}
```

**Comment by caffeine01** [comment]
Will this support things like CSS transforms? for example "-gtk-icon-transform" or just simply "transform".
For reference: https://docs.gtk.org/gtk4/css-properties.html

**Comment by tmccombs** [comment]
There is https://packages.debian.org/sid/libdbusmenu-gtk4

Although it might be debian specific.

It also might not be that bad to implement the menu manually instead of using a library.

**Comment by Rayzeq** [comment]
Although the package you linked is called `libdbusmenu-gtk4` for some reason, it's actually made for GTK-2+

**Comment by 4194304** [comment]
> I fully support this. Being able to not use things like blur effects for me isn't acceptable. For now I'll be using workarounds to that specific problem

Looking back at that is funny now since I use TWM, X11 cursors, hate GTK 4.0 (GTK theming), and am on my way to make my system completely monochrome (basically a classic X11 enthusiast)

**Comment by Beryesa** [comment]
> ...I use TWM, X11 cursors, hate GTK 4.0 (GTK theming), and am on my way to make my system completely monochrome...

This is only a GTK4 port, which has nothing to do with the theming. GTK4 is NOT just Libadwaita exclusively, they're indeed very separate things. You didn't even read up before having an opinion on it, let's not waste contributors' time here please.

**Comment by LukashonakV** [comment]
> There is https://packages.debian.org/sid/libdbusmenu-gtk4
> 
> Although it might be debian specific.
> 
> It also might not be that bad to implement the menu manually instead of using a library.
> 
> I'd be willing to help with implementing that.

Actually it's not. It's just like strange naming ...

**Comment by RobertMueller2** [comment]
Please note for this, I've added a change to the privacy module in PR #4092 . I think the behaviour with removing managed objects may change in GTK4, leaving the delete superfluous, at least that is a how I read https://gnome.pages.gitlab.gnome.org/gtkmm-documentation/chapter-memory.html compared to the behaviour observed in GTK3. I haven't spent much effort to really research this, so I might be mistaken. Just wanted to give a heads up.

**Comment by LukashonakV** [comment]
Hi @Alexays and everyone,
It seems like this issue has been stuck for a while, but in order to move forward, I would like to propose a slight change in approach.

1. @Alexays, is it possible to set up a new Waybar 4 repository?  
2. I have already created a similar one: https://github.com/LukashonakV/Waybar4. The goal is to deliver minor revisions under Major 4, with each minor revision introducing new functionality.  
3. Actually, I am starting to work mostly from scratch to align with the original codebase. Why? The existing source code looks chaotic, so I am not willing to adjust it this time. Instead, I plan to base the new code on what GTK4 proposes, rewriting it to utilize the features provided.  
4. For the new project, @Alexays, is it possible to set up an issue template? The original project has a long list of issues, many of which lack the necessary details to be actionable.  
5. @Alexays, you can update your master/main branch with the same from my repo each time a new version is released. I will actually stop doing it this way once we have some parity between repositories.  
6. I’m not going to support all these modules, but I will try to make at least some of them work under the new framework.  
7. The reason I changed my approach is that it's quite complex to try to bring new commits without the opportunity to incorporate necessary changes into the branch without the owner’s assistance.

Let me know if there are any objections. If there are, it’s not an issue for me to close the project.
Regards,Viktar Lukashonak

**Comment by LukashonakV** [comment]
Dear @Alexays,
Would it be possible for you to help set up the Waybar4 project, or create a 4-age branch, so we can start official tracking for the project using the GTK4 framework?

I’ve just made an initial commit that builds and runs the Waybar4 panel, but there aren’t any widgets yet. The codebase is fairly up to date with the original project. Please get codebase from the [Waybar4](https://github.com/LukashonakV/Waybar4)

**Comment by LukashonakV** [comment]
Hey @Alexays ,

I consider the Waybar migration to gtkmm4 finally complete. There are still a couple of loose ends that I think the community can finalize. This is my third — and final — attempt to help the project move to the new framework; I’ve already sunk quite a lot of time into it.

Please pull with -force flag the work from my project into the `gtk4` branch https://github.com/Alexays/Waybar/tree/gtk4 <- https://github.com/LukashonakV/Waybar/tree/gtk4_final. The current commit is [`https://github.com/LukashonakV/Waybar/commit/d0e83fe81d93395f93ca55444f1cd79b3352a413`](https://github.com/LukashonakV/Waybar/commit/d0e83fe81d93395f93ca55444f1cd79b3352a413).

----



All the best,
Viktar Lukashonak

**Comment by LukashonakV** [comment]
- Migrated the tray module (with significant help from Kimi)  
- Final commit: [`LukashonakV/Waybar@c22045f`](https://github.com/LukashonakV/Waybar/commit/c22045f4854cc6f9d93333700764c292bc2f8290)

----

@Alexays, heads up: new commits will be landing here periodically, including bug fixes.
